source/opt/interface_repair_pass.cpp

```diff
@@ -0,0 +1,78 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "interface_repair_pass.h"
+#include "source/spirv_constant.h"
+namespace spvtools {
+namespace opt {
+InterfaceRepairPass::Status InterfaceRepairPass::Process() {
+  std::unordered_map<uint32_t, Function*> id2function_;
+  for (auto& fn : *get_module()) id2function_[fn.result_id()] = &fn;
+  bool modified = false;
+  for (auto& entry : get_module()->entry_points()) {
+    std::set<uint32_t> used_variables;
+    std::function<void(Function * func)> traverser = [&](Function* func) {
+      for (auto bi = func->begin(); bi != func->end(); ++bi) {
+        for (auto ii = bi->begin(); ii != bi->end(); ++ii) {
+          if (ii->opcode() == SpvOpFunctionCall) {
+            const auto called_id = ii->GetOperand(2).words[0];
+            traverser(id2function_.at(called_id));
+          } else {
+            for (uint32_t i = 0; i < ii->NumOperands(); ++i) {
+              auto& op = ii->GetOperand(i);
+              if (op.type == SPV_OPERAND_TYPE_ID) {
+                auto id = op.words[0];
+                if (used_variables.count(id)) continue;
+                auto* var = get_def_use_mgr()->GetDef(id);
+                if (!var || var->opcode() != SpvOpVariable) continue;
+                auto storage_class = var->GetSingleWordInOperand(0);
+                if (storage_class != SpvStorageClassFunction &&
+                    (get_module()->version() >= SPV_SPIRV_VERSION_WORD(1, 4) ||
+                     storage_class == SpvStorageClassInput ||
+                     storage_class == SpvStorageClassOutput)) {
+                  used_variables.insert(id);
+                }
+              }
+            }
+          }
+        }
+      }
+    };
+    traverser(id2function_.at(entry.GetSingleWordInOperand(1)));
+    {
+      std::set<uint32_t> old_variables;
+      for (int i = entry.NumInOperands() - 1; i >= 3; --i) {
+        auto variable = entry.GetInOperand(i).words[0];
+        if (!used_variables.count(variable))  // There is unused ids.
+          goto repair;
+        if (old_variables.count(variable))  // There is duplicate ids.
+          goto repair;
+        old_variables.insert(variable);
+      }
+      if (old_variables.size() != used_variables.size())  // There is missing.
+        goto repair;
+    }
+    continue;
+  repair:
+    modified = true;
+    for (int i = entry.NumInOperands() - 1; i >= 3; --i)
+      entry.RemoveInOperand(i);
+    for (auto id : used_variables) {
+      entry.AddOperand(Operand(SPV_OPERAND_TYPE_ID, {id}));
```

Comment:
I'm getting the impression that this is going to remove the "input" and "output" variables as well, but I don't think it should.  That is part of the interface between stages that is suppose to match to a certain extent.  See this code in the [vulkan validation layers](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/20678fff2a19946fe8d5f3ee5300af1c3228268a/layers/shader_validation.cpp#L1906.).  By removing an input to the fragement shader, you could be causing a warning, which might not be all that bad.  However, by removing an output you could be causing an error.

Reply:
Hmm, the second case, I don't think the error is caused by this optimization, the original shader should not be valid.

For example: 
An output variable not written to by a stage is read from the next stage, this is definitely an **use-before-init BUG**.

If user did not apply this optmimization pass before linking, the variable is still included in the interface of both entry points.
Then, the validation layer can't detect such BUG.

If applied, the validation layer emits an error message.

Reply:
The first case, the validation layer emits an performance warning.
If an variable written to by an stage is not used by the next stage, computing such an useless variable is losing performance.

Reply:
I agree the performance warning can be a good thing.  However, in the other case, it is not necessarily a "use-before-init" bug.  The validation layers do a static analysis.  The input may be read statically, but not dynamically.  So I could see a situation where there are two VS that pair with one PS.  Both VS will declare the output, but only one writes to it.  The PS will read, but the read is guarded by a check that correlates with VS in the pipeline.  This is all legal because the user is being careful with what they are doing.  This pass will remove the output from the VS that does not write to it, and then there will be an error in the validation layers.

Reply:
But in Vulkan 1.2 specification, paragraph 15.1.3, "Interface Matching", there is: 
"Any input value to a shader stage is well-defined as long as the preceding stage writes to a matching output, as described above."
It does not mention whether the input variable is read by the subsequent stage or not, but it does say "write".
Maybe I am understanding wrongly about this.

Reply:
Did you notice the previous line:

```
At an interface between two shader stages, the user-defined variable interface must match exactly, as described above.
```

Technically, if the output is not written, it is reasonable to say that the matching input in the next stage is not well defined.  However, if it is not actually read, then it should not be a problem on real hardware.  I believe my scenario is still a good reason to not remove inputs and outputs in this pass.  The interface has to be declared by the original source and cannot be change without looking at the entire pipeline.

For the specific issue that is giving you a problem, that should be fixed in DXC.  It has the information to get the interface correct.  If we fix the problem in DXC do we still need this pass?

Reply:
I did notice the previous line, but I think if shaders are well-defined, the variable interfaces will match exactly although unused variables are removed by the pass.

From a game developer's perspective, this pass is still useful.

We often have many features in one 'uber-shader', we won't enable all features simultaneously, only a few of them, spirv-opt's DCE can remove code of disabled features at runtime. However varying variables used by disabled features should be removed too, number of varying variable components matters, which must below the 'maxFragmentInputComponents', on many hardware there is only 128, 128 is too limited for a big uber shader.

I think we can have this pass, it's not on by default, so people want to use it have the option.

Reply:
> I did notice the previous line, but I think for those 'well-defined' shaders, the variable interfaces will match exactly although unused variables are removed by the pass, because these variables are removed from both sides.

The scenario that I described was intended to show that the variable will not always be removed consistently:

```
void vert1() {
  shadowTexCoord = ........;
}

void vert2() {}

void frag() {
  if (paired_with_vert_1 /* the value is only known at run time */) {
    ... = shadowTexCoord;
  }
}
```

In any case, I can see a code style that could benefit from removing unused inputs and outputs; however it will have to be clear from the description that the user is responsible for meeting the criteria for making it work.  The pass should be renamed because it is not fixing a broken interface.

I would suggest a name like "remove-unused-interface-variables".  The description should be something like:

```
Removes variables referenced on the |OpEntryPoint| instruction that are not referenced in the entry point function or any function in its call tree.  Note that this could cause the shader interface to no longer match other shader stages.
```

Reply:
I agree that it should be renamed, but the pass also can add the missing interface variables if there is any, not only remove the unused variables, so maybe rename it like "redeclare-interface-variables" ?
I know that there shouldn't be any variables missing if the input spir-v is legal, so "remove-unused-interface-variables" may be better.

Reply:
Okay, I'll wait for an update that will change the name.

Reply:
I've made the change. In addition, is it necessary to squash my commits into one ?

Reply:
Thanks I'll take care of squashing the commits.